File: backend/src/main/java/com/example/backend/entity/RecruitPost.java
package com.example.backend.entity;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList; // Application 연관관계 위해 추가
import java.util.List;    // Application 연관관계 위해 추가
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import com.example.backend.enums.RecruitCategory;
import com.example.backend.enums.RecruitTargetType;
import com.example.backend.enums.ParticipantType;
import com.example.backend.enums.RecruitStatus;

@Entity
@Table(name = "recruit_posts")
@Getter @Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@ToString(exclude = {"author", "postingTeam", "applications"}) // 양방향 연관관계 필드 제외
@EqualsAndHashCode(exclude = {"author", "postingTeam", "applications"}) // 양방향 연관관계 필드 제외
public class RecruitPost {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", referencedColumnName = "id", nullable = false)
    private User author; // 게시글을 시스템에 등록한 사용자

    // ✅ 새로운 필드: 이 게시글의 주체가 되는 팀 (팀원/경기 모집 시)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "posting_team_id", referencedColumnName = "id", nullable = true)
    private Team postingTeam; // 팀이 주체일 경우 해당 팀, 개인이 주체면 null

    @Column(length = 255, nullable = false)
    private String title;

    @Lob // 긴 텍스트를 위해
    @Column(columnDefinition = "TEXT", nullable = false)
    private String content;

    @Column(length = 100, nullable = false)
    private String region;

    @Column(name = "sub_region", length = 100)
    private String subRegion;

    @Column(name = "thumbnail_url", length = 500)
    private String thumbnailUrl;

    @Enumerated(EnumType.STRING)
    @Column(length = 50, nullable = false)
    private RecruitCategory category; // 용병 모집인지 팀 모집인지 경기 모집인지를 결정하는 카테고리

    @Enumerated(EnumType.STRING)
    @Column(name = "target_type", length = 50, nullable = false)
    private RecruitTargetType targetType;

    @Enumerated(EnumType.STRING)
    @Column(name = "from_participant", length = 50, nullable = false)
    private ParticipantType fromParticipant;

    @Enumerated(EnumType.STRING)
    @Column(name = "to_participant", length = 50, nullable = false)
    private ParticipantType toParticipant;

    @Column(name = "game_date") // nullable = false 제거 (용병/팀원 모집 시 선택 사항일 수 있음)
    private LocalDate gameDate;

    @Column(name = "game_time") // nullable = false 제거
    private LocalTime gameTime;

    @Enumerated(EnumType.STRING)
    @Column(length = 50, nullable = false)
    private RecruitStatus status;

    @Column(name = "required_personnel")
    private Integer requiredPersonnel;

    @Column(name = "age_group", length = 50)
    private String ageGroup;

    @Column(name = "preferred_positions", length = 255)
    private String preferredPositions;

    @Lob
    @Column(name = "match_rules", columnDefinition = "TEXT")
    private String matchRules;

    @Column(name = "min_players")
    private Integer minPlayers;

    @Column(name = "max_players")
    private Integer maxPlayers;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    // 양방향 연관관계: 이 게시글에 대한 Application 목록
    @OneToMany(mappedBy = "recruitPost", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Application> applications = new ArrayList<>();
}

File: frontend/src/types/recruitPost.ts
// src/types/recruitPost.ts

export interface PostType { // API 응답 또는 화면 표시용 타입
  id: number;
  title: string;
  content: string;
  region: string;
  subRegion?: string;
  thumbnailUrl?: string; // PostType에는 있지만 CreationRequest에는 없을 수 있음, 또는 반대
  category: RecruitCategory; // Enum 값 사용
  targetType: RecruitTargetType; // Enum 값 사용
  fromParticipant: ParticipantType; // Enum 값 사용
  toParticipant: ParticipantType;   // Enum 값 사용
  gameDate?: string;
  gameTime?: string;
  status: RecruitStatus; // Enum 값 사용
  requiredPersonnel?: number;
  ageGroup?: string;
  preferredPositions?: string;
  matchRules?: string;
  minPlayers?: number;
  maxPlayers?: number;
  authorId: number | null;
  authorName: string | null;
  createdAt: string;
  updatedAt: string;
  // postingTeamId, postingTeamName 등 추가된 필드 고려
}

export enum RecruitCategory {
  MERCENARY = "MERCENARY",
  TEAM = "TEAM",
  MATCH = "MATCH",
}

export enum RecruitTargetType {
  USER = "USER",
  TEAM = "TEAM",
}

export enum ParticipantType {
  INDIVIDUAL = "INDIVIDUAL",
  TEAM = "TEAM",
}

export enum RecruitStatus {
  RECRUITING = "RECRUITING",
  COMPLETED = "COMPLETED",
  IN_PROGRESS = "IN_PROGRESS", // 백엔드 Enum에 정의되어 있다면
  FINISHED = "FINISHED",       // 백엔드 Enum에 정의되어 있다면
  CANCELLED = "CANCELLED",     // 백엔드 Enum에 정의되어 있다면
}

export enum ApplicationStatus {
  PENDING = 'PENDING',
  ACCEPTED = 'ACCEPTED',
  REJECTED = 'REJECTED',
}

/**
 * 게시글 생성 요청 시 서버로 전송할 데이터 타입.
 * 백엔드 RecruitPostCreationRequest.java 와 필드 구조 및 타입을 일치시킵니다.
 */
export interface RecruitPostCreationRequestDto {
  //authorId: number; // User 엔티티의 id 타입과 일치
  title: string;
  content: string;
  category: RecruitCategory; // Enum 값 전송
  region: string;
  subRegion?: string;
  gameDate?: string; // YYYY-MM-DD 형식의 문자열
  gameTime?: string; // HH:MM 형식의 문자열
  matchRules?: string;
  minPlayers?: number; // Java Integer -> number | undefined
  maxPlayers?: number; // Java Integer -> number | undefined
  ageGroup?: string;
  preferredPositions?: string;
  thumbnailUrl?: string;
  requiredPersonnel?: number;
  targetType: RecruitTargetType; // Enum 값 전송
  fromParticipant: ParticipantType; // Enum 값 전송
  toParticipant: ParticipantType;   // Enum 값 전송
  //status: RecruitStatus; // Enum 값 전송 (생성 시 기본값은 보통 RECRUITING)
}

/**
 * 백엔드로부터 받는 모집 게시글 응답 데이터 타입 (DTO)
 * GET /api/recruit-posts 또는 POST /api/recruit-posts 응답에 사용됩니다.
 */
export interface RecruitPostResponseDto {
  id: number;
  title: string;
  content: string;
  region: string;
  subRegion: string | null; // null일 수 있음
  thumbnailUrl: string | null;
  category: 'MERCENARY' | 'TEAM' | 'MATCH'; // Enum은 보통 문자열로 넘어옵니다.
  targetType: string; // 'TEAM_TO_INDIVIDUAL' | 'INDIVIDUAL_TO_TEAM' 등
  fromParticipant: string | null;
  toParticipant: string | null;
  gameDate: string; // "YYYY-MM-DD" 형식의 문자열
  gameTime: string; // "HH:mm:ss" 형식의 문자열
  status: 'RECRUITING' | 'COMPLETED'; // 모집중, 모집완료 등
  requiredPersonnel: number | null;
  ageGroup: string | null;
  preferredPositions: string | null;
  matchRules: string | null;
  minPlayers: number | null;
  maxPlayers: number | null;
  authorId: number | null;
  authorName: string | null;
  createdAt: string; // "YYYY-MM-DDTHH:mm:ss" 형식의 ISO 문자열
  updatedAt: string;
}


export type RecruitPostUpdateRequestDto = Partial<RecruitPostCreationRequestDto>;

